File: 01_Rust/src/worker.rs
use crate::incoming::behaviour::Behaviour;
use crate::incoming::work::Work;
use rand::rngs::StdRng;
use rand::{RngExt, SeedableRng};
use std::collections::HashSet;
use std::io;
use std::io::{Error, ErrorKind};
use std::sync::{Arc, Mutex as StdMutex, OnceLock};
use std::time::{Duration, Instant};
use tokio::sync::Mutex as TokioMutex;
use tokio_util::sync::CancellationToken;

use crate::incoming::execution_model::ExecutionModel;
use tokio::sync::mpsc::Receiver;
use tokio::time::sleep;

// Demo "exactly-once" ledger (process lifetime only).
static PROCESSED: OnceLock<StdMutex<HashSet<usize>>> = OnceLock::new();

fn create_ledger() -> &'static StdMutex<HashSet<usize>> {
    PROCESSED.get_or_init(|| StdMutex::new(HashSet::new()))
}

fn try_mark_processed(id: usize) -> bool {
    let mut map = create_ledger()
        .lock()
        .unwrap_or_else(|poisoned| poisoned.into_inner());
    map.insert(id)
}

pub struct Worker {
    receiver: Arc<TokioMutex<Receiver<Work>>>,
}

impl Worker {
    pub fn create(receiver: Arc<TokioMutex<Receiver<Work>>>) -> Worker {
        let _ = create_ledger();
        Worker { receiver }
    }

    pub async fn execute(&mut self, cancellation_token: CancellationToken) {
        loop {
            let maybe_work: Option<Work> = tokio::select! {
                _ = cancellation_token.cancelled() => {
                    println!("Worker cancelled");
                    return;
                }
                work = async {
                    let mut rx = self.receiver.lock().await;
                    rx.recv().await
                } => work,
            };

            let Some(work) = maybe_work else {
                // Channel closed and drained.
                return;
            };

            println!(
                "Working on message {} | Model={:?} | Behaviour={:?}",
                work.get_id(),
                work.get_execution_model(),
                work.get_behaviour(),
            );

            Self::simulate(&work, cancellation_token.clone()).await;

            println!("Done message {}", work.get_id());
        }
    }

    async fn simulate(work: &Work, cancellation_token: CancellationToken) {
        let behaviour = work.get_behaviour();

        // Exactly-once: skip duplicates (demo-level, in-   memory only).
        if behaviour.contains(Behaviour::EXACTLY_ONCE) && !try_mark_processed(work.get_id()) {
            println!(
                "  [ExactlyOnce] Work {} already processed -> skipping.",
                work.get_id()
            );
            return;
        }

        let attempts = if behaviour.contains(Behaviour::RETRYABLE) {
            3
        } else {
            1
        };

        for attempt in 1..=attempts {
            tokio::select! {
                _ = cancellation_token.cancelled() => {
                    println!("  Work cancelled");
                    return;
                }
                _ = async {
                    // HighPriority: less "queueing"/overhead before work begins.
                    if behaviour.contains(Behaviour::HIGH_PRIORITY) {
                        println!("  [HighPriority] Fast-lane execution.");
                    }

                    // RequiresAffinity: pin to a pretend partition/worker lane.
                    let lane = work.get_id() % 4;
                    if behaviour.contains(Behaviour::REQUIRES_AFFINITY) {
                        println!("  [RequiresAffinity] Routing to lane {}.", lane);
                    }

                    match Self::simulate_by_execution_model(work, lane).await {
                        Ok(_) => {
                            if behaviour.contains(Behaviour::RETRYABLE) {
                                println!("  [Retryable] Succeeded on attempt {}/{}.", attempt, attempts);
                            }
                            else {
                                println!("  Completed.");
                            };
                        }
                        Err(err) => {
                            println!("  Attempt {}/{} failed: {}", attempt, attempts, err);

                            if attempt == attempts {
                                println!("  Giving up.");
                                return;
                            }

                            // Tiny exponential-ish backoff for the demo.
                            let backoff_ms = 50 * attempt * attempt;
                             sleep(Duration::from_millis(backoff_ms)).await;
                        }
                    };
                } => {}
            }
        }
    }

    async fn simulate_by_execution_model(work: &Work, lane: usize) -> io::Result<()> {
        let behaviour = work.get_behaviour();

        // Base pacing. Behaviours can tweak it.
        let mut step_delay_ms = 80;

        if behaviour.contains(Behaviour::LONG_RUNNING) {
            step_delay_ms += 160;
            println!("  [LongRunning] Slower steps.");
        }

        if behaviour.contains(Behaviour::RESOURCE_INTENSIVE) {
            println!("  [ResourceIntensive] Adding CPU work.");
        }

        // Scheduled: pretend we had to wait until a trigger time.
        if work.get_execution_model() == ExecutionModel::Scheduled {
            println!("  [Scheduled] Waiting for trigger...");
            sleep(Duration::from_millis(150)).await;
        }

        match work.get_execution_model() {
            ExecutionModel::OneOff => {
                Self::step("OneOff: run once", behaviour, step_delay_ms).await?;
            }
            ExecutionModel::Scheduled => {
                Self::step("Scheduled: execute job", behaviour, step_delay_ms).await?;
            }
            ExecutionModel::EventDriven => {
                Self::step(
                    "EventDriven: handle event payload",
                    behaviour,
                    step_delay_ms,
                )
                .await?;
            }
            ExecutionModel::Batch => {
                println!("  [Batch] Processing items...");
                for i in 1..=5 {
                    Self::step(
                        format!("Batch item {}/5", i).as_str(),
                        behaviour,
                        step_delay_ms,
                    )
                    .await?;
                }
            }
            ExecutionModel::Stream => {
                println!("  [Stream] Polling/consuming stream ticks...");
                for i in 1..=4 {
                    Self::step(
                        format!("Stream tick {}/4", i).as_str(),
                        behaviour,
                        step_delay_ms + 40,
                    )
                    .await?;
                }
            }
            ExecutionModel::Workflow => {
                println!("  [Workflow] Running steps (DAG-ish)...");
                Self::step("Step A: validate", behaviour, step_delay_ms).await?;
                Self::step("Step B: transform", behaviour, step_delay_ms + 20).await?;
                Self::step("Step C: persist", behaviour, step_delay_ms + 40).await?;
            }
            ExecutionModel::Actor => {
                println!(
                    "  [Actor] Handling partition key lane={} (stateful-ish).",
                    lane
                );
                Self::step("Actor turn: load state", behaviour, step_delay_ms).await?;
                Self::step("Actor turn: apply work", behaviour, step_delay_ms + 20).await?;
                Self::step("Actor turn: save state", behaviour, step_delay_ms + 40).await?;
            }
        };

        Ok(())
    }

    async fn step(label: &str, behaviour: Behaviour, delay_ms: usize) -> io::Result<()> {
        let mut rng = StdRng::from_rng(&mut rand::rng());
        let mut maybe_fail_transiently = || {
            if !behaviour.contains(Behaviour::RETRYABLE) {
                return Ok(());
            }

            // Fail sometimes (more often when "resource intensive") so retries are visible.
            let odds = if behaviour.contains(Behaviour::RESOURCE_INTENSIVE) {
                4
            } else {
                7
            }; // 1/4 or 1/7

            if rng.random_range(0..odds) == 0 {
                return Err(Error::new(
                    ErrorKind::Interrupted,
                    "Transient failure (simulated).",
                ));
            }

            Ok(())
        };

        let cpu_bump = |iterations: usize| {
            // ResourceIntensive: do a tiny CPU spin per step (demo only).
            let start = Instant::now();

            let mut x: u64 = 0;
            for i in 0..iterations as u64 {
                x = x.wrapping_mul(31) ^ i;
            }

            std::hint::black_box(x);

            let _elapsed = start.elapsed();
        };

        println!("  -> {}", label);

        if behaviour.contains(Behaviour::RESOURCE_INTENSIVE) {
            cpu_bump(25_000);
        }

        match maybe_fail_transiently() {
            Ok(_) => {
                sleep(Duration::from_millis(delay_ms as u64)).await;
                Ok(())
            }
            Err(err) => Err(err),
        }
    }
}

File: 06_Kotlin/workers/src/main/kotlin/Worker.kt
package org.andreastme.langpermonth

import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.delay
import org.andreastme.langpermonth.incoming.Behaviour
import org.andreastme.langpermonth.incoming.ExecutionModel
import org.andreastme.langpermonth.incoming.Work
import java.util.concurrent.ConcurrentHashMap
import kotlin.math.abs
import kotlin.random.Random
import kotlin.time.Duration.Companion.milliseconds

// Demo "exactly-once" ledger (process lifetime only).
private val Processed = ConcurrentHashMap<Int, Boolean>()

class Worker(
    private val channel: Channel<Work>
) {
    suspend fun execute() {
        while (true) {
            val result = channel.tryReceive()
            if (result.isClosed) break

            val work = result.getOrNull() ?: continue
            println("Working on item ${work.id} | Model=${work.executionModel.name} | Behaviour=${work.behaviour}")

            simulate(work)

            println("Done work ${work.id}.")
        }
    }

    private suspend fun simulate(work: Work) {
        // Exactly-once: skip duplicates (demo-level, in-memory only).
        if (work.behaviour.hasFlag(Behaviour.ExactlyOnce) && Processed.containsKey(work.id)) {
            println("  [ExactlyOnce] Work ${work.id} already processed -> skipping.")
            return
        }

        Processed[work.id] = true

        val attempts = if (work.behaviour.hasFlag(Behaviour.Retryable)) 3 else 1

        for (attempt in 1..attempts) {
            try {
                // HighPriority: less "queueing"/overhead before work begins.
                if (work.behaviour.hasFlag(Behaviour.HighPriority)) {
                    println("  [HighPriority] Fast-lane execution.")
                }

                // RequiresAffinity: pin to a pretend partition/worker lane.
                val lane = abs(work.id) % 4
                if (work.behaviour.hasFlag(Behaviour.RequiresAffinity)) {
                    println("  [RequiresAffinity] Routing to lane $lane.")
                }

                simulateByExecutionModel(work, lane)

                println(
                    if (work.behaviour.hasFlag(Behaviour.Retryable)) {
                        "  [Retryable] Succeeded on attempt $attempt/$attempts."
                    } else {
                        "  Completed."
                    }
                )

                return
            } catch (e: Throwable) {
                if (e is CancellationException) {
                    println("  Work cancelled")
                    throw e
                }

                println("  Attempt $attempt/$attempts failed: ${e.message}")

                if (attempt == attempts) {
                    println("  Giving up.")
                    return
                }

                // Tiny exponential-ish backoff for the demo.
                val backoffMs = 50 * attempt * attempt
                delay(backoffMs.milliseconds)
            }
        }
    }

    private suspend fun simulateByExecutionModel(work: Work, lane: Int) {
        // Base pacing. Behaviours can tweak it.
        var stepDelayMs = 80

        if (work.behaviour.hasFlag(Behaviour.LongRunning)) {
            stepDelayMs += 160
            println("  [LongRunning] Slower steps.")
        }

        if (work.behaviour.hasFlag(Behaviour.ResourceIntensive)) {
            println("  [ResourceIntensive] Adding CPU work.")
        }

        // Scheduled: pretend we had to wait until a trigger time.
        if (work.executionModel == ExecutionModel.Scheduled) {
            println("  [Scheduled] Waiting for trigger...")
            delay(150.milliseconds)
        }

        when (work.executionModel) {
            ExecutionModel.OneOff -> {
                step(work, "OneOff: run once", stepDelayMs)
            }

            ExecutionModel.EventDriven -> {
                step(work, "EventDriven: handle event payload", stepDelayMs)
            }

            ExecutionModel.Batch -> {
                println("  [Batch] Processing items...")
                for (i in 1..5) {
                    step(work, "Batch item $i/5", stepDelayMs)
                }
            }

            ExecutionModel.Stream -> {
                println("  [Stream] Polling/consuming stream ticks...")
                for (tick in 1..4) {
                    step(work, $"Stream tick $tick/4", stepDelayMs + 40)
                }
            }

            ExecutionModel.Workflow -> {
                println("  [Workflow] Running steps (DAG-ish)...")
                step(work, "Step A: validate", stepDelayMs)
                step(work, "Step B: transform", stepDelayMs + 20)
                step(work, "Step C: persist", stepDelayMs + 40)
            }

            ExecutionModel.Actor -> {
                println("  [Actor] Handling partition key lane=$lane (stateful-ish).")
                step(work, "Actor turn: load state", stepDelayMs)
                step(work, "Actor turn: apply work", stepDelayMs + 20)
                step(work, "Actor turn: save state", stepDelayMs + 40)
            }

            ExecutionModel.Scheduled -> {
                step(work, "Scheduled: execute job", stepDelayMs)
            }
        }
    }

    // Random transient failure for Retryable demos.
    private fun maybeFailTransiently(work: Work) {
        if (!work.behaviour.hasFlag(Behaviour.Retryable)) {
            return
        }

        // Fail sometimes (more often when "resource intensive") so retries are visible.
        val odds = if (work.behaviour.hasFlag(Behaviour.ResourceIntensive)) 4 else 7 // 1/4 or 1/7
        if (Random.nextInt(odds) == 0) {
            throw IllegalStateException("Transient failure (simulated).")
        }
    }

    // ResourceIntensive: do a tiny CPU spin per step (demo only).
    @Suppress("SameParameterValue")
    private fun cpuBump(iterations: Int) {
        var x = 0

        for (i in 0 until iterations) {
            x = (x * 31) xor i
        }

        x.hashCode()
    }

    private suspend fun step(work: Work, label: String, delayMs: Int) {
        println("  -> $label")
        if (work.behaviour.hasFlag(Behaviour.ResourceIntensive)) {
            cpuBump(25_000)
        }

        maybeFailTransiently(work)
        delay(delayMs.milliseconds)
    }
}
